### ISSUE TITLE
Change Request: export rules for reuse

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#19013]
### ESLint version

v9.12.0

### What problem do you want to solve?

See https://github.com/eslint/eslint/discussions/15011 for more context.

In short, I want to be able to declare and reuse a core rule with multiple settings. For example:

```js
{
	plugins: {
		fantozzi: {
			rules: {
				'no-lodash': require('eslint/lib/rules/no-restricted-imports'),
				'prefer-preact': require('eslint/lib/rules/no-restricted-imports'),
			},
		},
	},
	rules: {
		'fantozzi:no-lodash': [
			'error',
			{
				'patterns': [
					{
						'message': 'No lodash plz',
						'paths': [
							'lodash',
							'underscore'
						]
					},
				]
			}
		],
		'fantozzi:prefer-preact': [
			'error',
			{
				'patterns': [
					{
						'message': 'Use preact',
						'paths': [
							'react'
						]
					},
				]
			}
		],
	}
}
```

This for example allows shared configs to expose a more granular set of rules, and allow consumers to disable specific parts of the rule.

### What do you think is the correct solution?

I think this is currently gate-kept by the `exports` field, which should be broadened to allow imports of specific rules:

https://github.com/eslint/eslint/blob/7dd402d33226d821a17b22c4753ce9c40fc909bd/package.json#L21-L23

Also some documentation wouldn't hurt, following what was mentioned in https://github.com/eslint/eslint/issues/15620#issuecomment-1048388514

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
You can access core rules through `eslint/use-at-your-own-risk` entry point:

```js
const { builtinRules } = require("eslint/use-at-your-own-risk");

const noRestrictedImports = builtinRules.get("no-restricted-imports");
```

Does this work for your use case?

**Comment by fregante** [comment]
The name implies that it's not a stable API so it's not fit to be safely used in shared configs. That would be the important part.

**Comment by aarongoldenthal** [comment]
> It's in unstable APIs because rule APIs are indeed not stable. Some custom rules inherit core rules and override visitors and/or reporting errors, which can break in a minor or patch version if we make any changes or even just refactor the rule. Otherwise, if a rule is used as-is, it should be safe.
> 
> @eslint/eslint-tsc thoughts about this, should we export rules in another way to support this use case?

It (`builtinRules`) is also flagged as deprecated [here](https://github.com/eslint/eslint/blob/1def4cdfab1f067c5089df8b36242cdf912b0eb6/lib/types/use-at-your-own-risk.d.ts#L31).

**Comment by aarongoldenthal** [comment]
When trying to use this as a way to set core rules with two settings and getting some errors, one of them was `Could not find "complexity" in plugin "core". Did you mean "@/complexity"?`. So, I tried this config and it appears without an export the core rules are also available under `@/`:

`eslint.config.js`:

```js
'use strict';

const complexityThresholdWarn = 1;
const complexityThresholdError = 2;

module.exports = [
    {
        files: ['**/*.{js,mjs,cjs}'],
        languageOptions: {
            sourceType: 'commonjs'
        },
        rules: {
            complexity: ['warn', complexityThresholdWarn],
            '@/complexity': ['error', complexityThresholdError]
        }
    }
];
```

Run against a variant of the `complexity` example results in:

```sh
E:\foo.js
  3:1  warning  Function 'a' has a complexity of 6. Maximum allowed is 1  complexity
  3:1  error    Function 'a' has a complexity of 6. Maximum allowed is 2  @/complexity
```

**Comment by fasttime** [comment]
I think built-in rules should be moved into the `@eslint/js` plugin and should be exported from there along with the `js` language definition. But that will be more than just refactoring because built-in rules are handled in a special way by the core and also JavaScript is the default language.

**Comment by nzakas** [comment]
For now, we shouldn't make any changes to how we are exporting core rules. Core rules are *not* intended to extended any you do so at your own risk, which is why `builtinRules` is exported from `/use-at-your-own-risk`.

In the future, the rules will move into the `@eslint/js` plugin, at which point you'll be able to access them through the `rules` key, but as @fasttime stated, that's a fairly major refactor that is not likely to be completed soon.

I think our options here are 1) leave this open as a marker for the `@eslint/js` work, or 2) close this as wontfix.

**Comment by mdjermanovic** [comment]
> 1) leave this open as a marker for the `@eslint/js` work

I think we don't have a tracking issue for the `@eslint/js` work yet, so it seems fine to leave this open as a marker. If and when we open an issue for the `@eslint/js` work where moving rules is listed as a task, we could link to it and close this.

**Comment by nzakas** [comment]
Okay, moving this to "Blocked" and adding to the v10 board to remind us to discuss in that timeframe.

**Comment by aladdin-add** [comment]
I pushed a small package for this: https://www.npmjs.com/package/@weiran.zsd/multi-eslint-rule-config

**Comment by sam3k** [comment]
The TSC group has decided to defer this as it would delay `v10`.

**Comment by jakubmazanec** [comment]
I have another use case for exporting the core rules, in case you're thinking about removing the currently deprecated export in the future. I prefer not to use any presets (I'm my experience ESLint plugins don't follow SemVer correctly and introduce rules set to `error` in minor versions) and instead I have all rules specified in my config even if they're off. And I have created [script](https://github.com/jakubmazanec/tools/blob/development/packages/eslint-config/tools/check-rules/main.ts) that checks all rules defined by ESLint and by plugins I use and shows me if there's a new rule I don't have in my config of if a rule was deprecated.

**Comment by fregante** [comment]
I published a tiny plugin for my use case (`no-restricted-syntax`)

- https://github.com/fregante/eslint-plugin-byo
